### PR TITLE
logging: some simplifications/cleanups

### PR DESCRIPTION
**Pull request by bluetech**
While looking at optimizing the logging plugin (since it shows up in profiles), and also at some issues and open PRs, I collected some simplifications and cleanups. I think they make the file a little cleaner and easier to understand and modify, and also a little faster, though this isn't the goal of this PR.

<details>
<summary>Benchmark</summary>

On this benchmark

```py
import pytest
@pytest.mark.parametrize("x", range(5000))
def test_foo(x): pass
```

Before:

    ============================ 5000 passed in 13.66s =============================
            13345882 function calls (12431103 primitive calls) in 14.054 seconds

After:

    ============================ 5000 passed in 12.53s =============================
            11271238 function calls (10716458 primitive calls) in 12.938 seconds
</details>

It is better to review each commit separately -- one of the commits reorders some methods so the PR diff will be unreadable.

Includes a few breaking changes to private attributes, and one deprecation removal (`--no-print-logs`) that was planned for 6.0.

This will cause conflicts with some open PRs. This is intentional, since I think they will be easier to evaluate after these changes (and some others). I will comment on them separately once it becomes relevant.

### CODE REVIEW

**Comment by bluetech** [comment]
Eek, looks like there is some issue on Python<=3.6. I'll take a look later.

File: src/_pytest/logging.py [hunk 1]
```diff
@@ -614,27 +614,17 @@ def pytest_collection(self) -> Generator[None, None, None]:
...
-            try:
-                yield
-            finally:
-                if item is not None and when == "teardown":
-                    del item._store[catch_log_handlers_key]
-                    del item._store[catch_log_handler_key]
+            item._store[catch_log_handlers_key][when] = log_handler
+            item._store[catch_log_handler_key] = log_handler
+
+            yield
```
Comment (nicoddemus) [inline src/_pytest/logging.py:638]: `try/finally` probably required here?

File: changelog/TBD.breaking.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+The `item.catch_log_handler` and `item.catch_log_handlers` attributes, set by the
+logging plugin, are no longer available.
```
Comment (nicoddemus) [inline changelog/TBD.breaking.rst:2]:
```suggestion
logging plugin and never meant to be public, are no longer available.
```

File: src/_pytest/logging.py (hunk 1 above)
Comment (bluetech) [reply src/_pytest/logging.py:638]: The `yield` can't throw (pluggy only `send`s a Result in), so it's not really needed.

File: changelog/TBD2.breaking.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+The deprecated ``--no-print-logs`` option is removed. Use ``--show-capture`` instead.
```
[the remark this answers is not in the document]
Comment (bluetech) [reply changelog/TBD2.breaking.rst:1]: The README.rst the `changelog` directory says to use `breaking` for removals, and doesn't mention `removal`. Should I use `removal` nonetheless?

[the remark this answers is not in the document]
Comment (nicoddemus) [reply changelog/TBD2.breaking.rst:1]: Ahh sorry, brainfart on my part: we had "removal" in the past, but then changed to "breaking" afterwards. 👍 

So nevermind. 😬

**Pull request merged**